2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/activity/search/SearchViewModel.kt

```diff
@@ -36,25 +72,33 @@ class SearchViewModel(private val naverRepository: NaverRepositoryImpl) {
 
     private fun onSuccess(movieList: List<MovieModel>) {
         if (movieList.isNotEmpty()) {
-            this.movieList.clear()
-            this.movieList.addAll(movieList)
+            _movieList.value?.also {
+                it.clear()
+                it.addAll(movieList)
+                _movieList.value = it
+            }
         } else {
-            showMessageEmptyResult.notifyChange()
+            _toastMessage.value = resourceProvider.getString(R.string.not_found_result)
         }
-        showProgressBar(false)
+        setVisibleProgressBar(false)
     }
 
     private fun onFailure(t: Throwable) {
         Log.d("chul", "OnFailure : $t")
-        showProgressBar(false)
+        setVisibleProgressBar(false)
     }
 
-    private fun showProgressBar(visible: Boolean) {
-        isLoading.set(visible)
+    private fun setVisibleProgressBar(visible: Boolean) {
+        _isLoading.value = visible
     }
 
+
     private fun clearCacheData(keyword: String) {
         naverRepository.clearCacheData()
     }
 
+    fun unbindViewModel() {
+        compositeDisposable.clear()
+    }
+
```

Comment:
```suggestion
    override fun onCleared() {
        compositeDisposable.clear()
        super.onCleared()
    }
```
Activity에서 control하는 것(1) 보다는 activity는 disposable을 모르는게 좋을 것 같고 viewmodel 생명주기 관련은 viewmodel 안에서 작성하는게 좋을 것 같습니다.(2)
구글링 해보면  
`ViewModel은 액티비티 스코프가 완전히 종료되는 시점에 종료가 되고, 이때 onCleared() 함수가 호출됩니다. 이 콜백은 ViewModel 클래스의 유일한 함수이며 ViewModel의 리소스를 해제하기에 적합한 곳입니다.` 
라고 나와있습니다.

Reply:
의견 감사합니다. 
onCleared()는 Activity의 onDestory 다음에 오게 되네요.
onCreate - onCleared의 생명주기보다는 onResume - onPause의 생명주기에서 Disposable을 clear하려는 의도였습니다!

Reply:
제 말은 ViewModel클래스에서 생명주기에 따라서 리소스 해제하도록 onCleard를 제공하는 이유가 있을텐데 굳이 왜 onResume, onPause 생명주기를 이용하셔서 불완전하게 하시는건지 궁금합니다.  단순히 그냥 그때 하고 싶으신건지요? 아니면 개발할 때 onCleard를 이용할떄보다 더 좋은 이점이 있는건가요?

Reply:
위에 코멘트 주시면 내용이랑 함께 의견을 적겠습니다.

> 제가 말씀드리고 싶은건 ViewModel클래스에서 생명주기에 따라서 리소스 해제하도록 onCleard를 제공하는 이유가 있을텐데 굳이 왜 onResume, onPause 생명주기를 이용하셔서 불완전하게 하시는건지 궁금합니다. 단순히 그냥 그때 하고 싶으신건지요? 아니면 개발할 때 onCleard를 이용할떄보다 더 좋은 이점이 있는건가요?

 예를 들겠습니다.
ViewModel에서 10초 걸리는 네트워크 통신 이후에 A전환이 일어나는 로직이 Rx Java로 동작하고 있다고 가정해보겠습니다. 그런데 사용자가 3초만에 홈키를 눌러서 앱을 onPause로 전환 시켰습니다. 그럴경우 어떻게 될까요? 10초 걸리는 작업의 Dispose가 Clear 되지 않았기 때문에 10초후에 화면전환이 일어나면서 앱이 올라오겠죠. 사용자의 관점에서는 홈키를 눌렀으니 코드 동작이 멈추는것을 기대했을거라 생각합니다. 그래서 onResume - onPause의 시점에서 Disposable을 Clear했습니다.

> 그리고 두번째로 onResume을 타기 전에 사용자가 어떠한 동작을 하면 앱은 어떻게 되나요 아니면 사용자가 항상 onResume후 vm세팅 후에 앱을 동작할거라는 보장이 있는지 궁금합니다.

onResume을 타기전에 사용자가 어떠한 동작을 할 수 있는지 잘 모르겠습니다. onResume을 지나서야 사용자와 상호작용을 할 수 있는것으로 알고 있습니다. 또한 이전 코멘트에서의 안전하지 않다는 의도를 잘 모르겠습니다.

Reply:
onStart관련해서 말씀드렸습니다. 그런 의도로 하신거군요 배웠습니다. 감사합니다

Reply:
홈키를 눌러서 이전 동작을 취소하는거라면 맞는 동작입니다.